plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,307 @@
+        return ERR;
+    }
+
+    return OK;
+}
+
+void AudioSourceM4A::close() throw() {
+    if (m_hDecoder) {
+        NeAACDecClose(m_hDecoder);
+        m_hDecoder = NULL;
+    }
+    if (MP4_INVALID_FILE_HANDLE != m_hFile) {
+        MP4Close(m_hFile);
+        m_hFile = MP4_INVALID_FILE_HANDLE;
+    }
+    m_trackId = MP4_INVALID_TRACK_ID;
+    m_maxSampleId = MP4_INVALID_SAMPLE_ID;
+    m_curSampleId = MP4_INVALID_SAMPLE_ID;
+    m_inputBuffer.clear();
+    m_inputBufferOffset = 0;
+    m_inputBufferLength = 0;
+    m_curFrameIndex = 0;
+    Super::reset();
+}
+
+bool AudioSourceM4A::isValidSampleId(MP4SampleId sampleId) const {
+    return (sampleId >= kMinSampleId) && (sampleId <= m_maxSampleId);
+}
+
+AudioSource::diff_type AudioSourceM4A::seekFrame(diff_type frameIndex) {
+    if (m_curFrameIndex != frameIndex) {
+        const MP4SampleId sampleId = kMinSampleId
+                + (frameIndex / kFramesPerSampleBlock);
+        if (!isValidSampleId(sampleId)) {
+            return m_curFrameIndex;
+        }
+        if ((m_curSampleId != sampleId) || (frameIndex < m_curFrameIndex)) {
+            // Restart decoding one or more blocks of samples
+            // backwards to avoid audible glitches.
+            // Implementation note: The type MP4SampleId is unsigned so we
+            // have to be careful when subtracting!
+            if ((kMinSampleId + kSampleIdPrefetchCount) < sampleId) {
+                m_curSampleId = sampleId - kSampleIdPrefetchCount;
+            } else {
+                m_curSampleId = kMinSampleId;
+            }
+            m_curFrameIndex = (m_curSampleId - kMinSampleId) * kFramesPerSampleBlock;
+            // rryan 9/2009 -- the documentation is sketchy on this, but I think that
+            // it tells the decoder that you are seeking so it should flush its state
+            faacDecPostSeekReset(m_hDecoder, m_curSampleId);
+            // discard input buffer
+            m_inputBufferOffset = 0;
+            m_inputBufferLength = 0;
+        }
+        // decoding starts before the actual target position
+        DEBUG_ASSERT(m_curFrameIndex <= frameIndex);
+        const size_type prefetchFrameCount = frameIndex - m_curFrameIndex;
+        // prefetch (decode and discard) all samples up to the target position
+        DEBUG_ASSERT(frames2samples(prefetchFrameCount) <= m_prefetchSampleBuffer.size());
+        readFrameSamplesInterleaved(prefetchFrameCount, &m_prefetchSampleBuffer[0]);
+    }
+    DEBUG_ASSERT(m_curFrameIndex == frameIndex);
+    return frameIndex;
+}
+
+AudioSource::size_type AudioSourceM4A::readFrameSamplesInterleaved(
+        size_type frameCount, sample_type* sampleBuffer) {
+    sample_type* pSampleBuffer = sampleBuffer;
+    const diff_type readFrameIndex = m_curFrameIndex;
+    size_type readFrameCount = m_curFrameIndex - readFrameIndex;
+    while ((readFrameCount = m_curFrameIndex - readFrameIndex) < frameCount) {
```

Comment:
This one is hard to read. Do you see a chance to move the assignment out of this while condition?